File: backend/src/components/accounts/entities/ledger_head_entity.py
from sqlalchemy import Column, String, ForeignKey, Enum
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from src.components.accounts.ledger_head_type_enum import LedgerHeadType
from src.utils.base_entity import BaseEntity


class LedgerHead(BaseEntity):
    __tablename__ = "ledgerheads"

    name = Column(String, nullable=False)
    code = Column(String, nullable=False)
    ledger_head_type = Column(Enum(LedgerHeadType, name="ledgerheadtype_enum", create_constraint=True, native_enum=False),
                  nullable=False)
    parent_id = Column(UUID(as_uuid=True), ForeignKey("ledgerheads.id"), nullable=False)

    lines = relationship("JournalLine", back_populates="ledgerhead", cascade="all, delete-orphan",
                           order_by="LedgerHead.created_date.desc()")

File: nodeapi/src/accounts/entities/ledger_head.entity.ts
import {
  Entity,
  Column,
  ManyToOne,
  OneToMany,
  JoinColumn,
} from 'typeorm';

import { BaseEntity } from 'src/common/entities/base.entity';
import { LedgerHeadType } from '../types/ledger_head_types.enum';
import { JournalLine } from './journal_lines.entity';
import { AccountKey } from '../services/transaction_rules.service';

@Entity('ledgerheads')
export class LedgerHead extends BaseEntity {
  @Column({
    type: 'varchar',
    nullable: false,
  })
  name: string;

  @Column({
    type: 'varchar',
    nullable: false,
  })
  code: string;

  @Column({
    type: 'enum',
    enum: LedgerHeadType,
    enumName: 'ledgerheadtype_enum',
    nullable: false,
  })
  ledgerHeadType: LedgerHeadType;

  @Column({
    type: 'uuid',
    name: 'parent_id',
    nullable: true,
  })
  parentId: string;

  @ManyToOne(
    () => LedgerHead,
    (ledgerHead) => ledgerHead.children
  )
  @JoinColumn({ name: 'parent_id' })
  parent: LedgerHead | null;

  @OneToMany(
    () => LedgerHead,
    (ledgerHead) => ledgerHead.parent
  )
  children: LedgerHead[];

  @OneToMany(
    () => JournalLine,
    (line) => line.ledgerHead,
    {
      cascade: true,
    }
  )
  lines: JournalLine[];

  @Column({
    type: 'enum',
    enum: AccountKey,
    enumName: 'accountkey_enum',
    nullable: false,
  })
  accountKey: AccountKey;
}